File: Cpp/962_Maximum_Width_Ramp.cpp
class Solution {
public:
    int maxWidthRamp(vector<int>& nums) {
        int result = 0;
        int n=nums.size();
        vector<int> left(n, nums[0]);
        vector<int> right(n, nums[n-1]);
        
        for(int i=1; i<n; ++i) {
            left[i] = min(left[i-1], nums[i]);
        }

        for(int i=n-2; i>=0; --i) {
            right[i]=max(right[i+1], nums[i]);
        }
        
        int i=0, j=0;
        while(j<n) {
            if(left[i] <= right[j]) {
                result = max(result, j-i);
                ++j;

            } else {
                ++i;
            }
        }
        
        return result;
    }
};

File: Kotlin/962_Maximum_Width_Ramp.kt
class Solution {
    fun maxWidthRamp(nums: IntArray): Int {
        var result = 0
        val n = nums.size
        val left = IntArray(n) { nums[0] }
        val right = IntArray(n) {nums[n - 1] }
        
        for(i in 1 until n) {
            left[i] = minOf(left[i - 1], nums[i])
        }
        
        for(i in n - 2 downTo 0) {
            right[i] = maxOf(right[i + 1], nums[i])
        }
        
        var i = 0
        var j = 0
        while(j < n) {
            if(left[i] <= right[j]) {
                result = maxOf(result, j-i)
                ++j
            } else {
                ++i
            }
        }
        
        return result
    }
}